### DISCUSSION TITLE
FASTAPI database connection pool

### DISCUSSION THREAD

**Discussion by MrYZhou** [fastapi/fastapi#10874] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
db.py
import os

from dotenv import load_dotenv
from sqlmodel import SQLModel, create_engine
import records


load_dotenv()
DB_HOST = os.getenv("DB_HOST") if os.getenv("DB_HOST") else "127.0.0.1"
DB_PORT = os.getenv("DB_PORT") if os.getenv("DB_PORT") else "3306"
DB_USER = os.getenv("DB_USER") if os.getenv("DB_USER") else "root"
DB_PASSWORD = os.getenv("DB_PASSWORD") if os.getenv("DB_PASSWORD") else "123456"
DB_NAME = os.getenv("DB_NAME") if os.getenv("DB_NAME") else "study"
DB_DRIVER = os.getenv("DB_DRIVER") if os.getenv("DB_DRIVER") else "mysql+pymysql"
SQLMODEL_ECHO = bool(os.getenv("SQLMODEL_ECHO"))
DB_HOST = os.getenv("DB_HOST") if os.getenv("DB_HOST") else "127.0.0.1"
DB_PORT = os.getenv("DB_PORT") if os.getenv("DB_PORT") else "3306"

DB_URL = f"{DB_DRIVER}://{DB_USER}:{DB_PASSWORD}@{DB_HOST}:{DB_PORT}/{DB_NAME}?charset=utf8mb4"
print(f"connect url : {DB_URL}")
print(f"print sql  : {SQLMODEL_ECHO}")
engine = create_engine(DB_URL, echo=SQLMODEL_ECHO)

system.py
from db import engine
xxx

connect.py 
from db import engine
xxxx
```

### Description

How to make a connection pool for fastapi, according to the official practice, every file must create a new connection. I want to share the engine

The current behavior is to go through the process of creating a connection every time an import is made, and I hope to return it directly

### Operating System

Windows

### FastAPI Version

latest ,0.95.1

### Pydantic Version

1.10.7

### Python Version

Python 3.11.2

**Comment by jgould22** [comment, marked as answer]
Something like this 
```python
settings = get_settings()

# Init the engine
engine = create_async_engine(
    settings.db_connection_string,
)

# Create a Session
async_session_factory = sessionmaker(
    engine,
    expire_on_commit=False,
    class_=_AsyncSession,
    autocommit=False,
)


async def get_session() -> _AsyncSession:
    """
    Yield db session
    """

    async with async_session_factory() as session:
        try:
            yield session
            await session.commit()
        except SQLAlchemyError as sql_ex:
            await session.rollback()
            raise sql_ex
        finally:
            await session.close()
```

and then something like what is described here https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#a-database-dependency-with-yield

The process is roughly the same if you are using a synchronous client like  `mysql+pymysql` just make sure to only use it in def endpoints so you do not block the event loop
